### PR TITLE
A path to avoiding dataset regressions 

### PR DESCRIPTION
**Pull request by anuragprat1k**
# Description

**This PR attempts to fix the following problems**:
1. The droidlet repo has no control over the dataset / models it depends on - any underlying change in dataset / models is thus unvetted and can cause failures in the repo. We have seen this happen on multiple occasions. 
2. Dataset / model uploads are not tested through CI.
3. There is no notion of versioning for the files uploaded to S3. There is only ever one datasets or models tar file on S3. 

**Fixes**:
* **To fix Problem 1** - I moved the checksums to the repo instead of on S3. These are the checksums that the repo now uses as defaults and also controls them since they are checked into source control (they live in `tools/data_scripts/default_checksums/`, they don't live on S3 anymore). **To move to using a new dataset / model, these hashes need to be updated through a pull-request (hence also fixing Problem 2 - lack of a CI run when moving to new assets).**
* **To fix Problem 3** - Datasets are now uploaded as `dataset_folder_<sha1sum>.tar.gz`, and the `sha1sum` acts as the version. The data scripts only download the tarball corresponding to the default checksums in `tools/data_scripts/default_checksums/`.

**New flow for uploading datasets**
1. After making dataset changes, use `tools/data_scripts/compute_checksum.sh` to save new checksums to `tools/data_scripts/default_checksums/`.
image

2. Upload dataset to s3 using `tools/data_scripts/tar_and_hash_datasets.sh`
image

3. Create a PR with changes to `tools/data_scripts/default_checksums/` from step 1. This PR checks for regressions in the new datasets folder. **Everyone moves to the new datasets folder only when this PR is merged.**

**Some other cosmetic changes**
* Moved the checksum calculation to one file (checksum_fn.sh) that every other file sources. This is to centralize the checksum logic rather than having multiple pieces of it. Fixed it to also use relative paths for calculating checkums (was using the entire path before and hence the same datasets folder in locobot and craftassist would give different checksums)
* Because there are no checksums on S3 now, I got rid of the file `tools/data_scripts/compute_and_upload_checksum.sh`.
* Renamed the file `compare_directory_hash.sh` to `try_download.sh` to capture more accurately what it does - tries downloads by comparing hashes.
* Fixed the hash computation to remove trailing hyphens and whitespaces.  
* Added a prompt before uploading to S3 in `tools/data_scripts/tar_and_hash_datasets.sh`. 

**Follow-ups**
* Model uploads are not versioned. We can do that along the same lines.
* Find a way to prune older datasets, since there will be multiple versions after this.
* Move away from bash entirely perhaps? The logic is getting messy. 

Fixes # (issue)

## Type of change

Please check the options that are relevant.

- [ ] Bug fix (non-breaking change that fixes an issue)
- [ ] Proposes a change (non-breaking change that isn't necessarily a bug)
- [ ] Refactor
- [x] New feature (non-breaking change that adds a new functionality)
- [ ] Breaking change (fix or feature that would break some existing functionality downstream)
- [ ] This is a unit test
- [ ] Documentation only change

## Type of requested review

- [x] I want a thorough review of the implementation.
- [ ] I want a high level review. 
- [ ] I want a deep design review.

# Testing
**Scenario 1: Clean downloads locobot**
1. `./try_download.sh locobot` - downloads models and datasets.
2. `./try_download.sh locobot` - doesn't download anything since assets exist.

**Scenario 2: Clean downloads craftassist**
1. `./try_download.sh craftassist` - downloads models and datasets.
2. `./try_download.sh craftassist` - doesn't download anything since assets exist.

**Scenario 3: Tested running the agent as well to check that files download on the first try and don't redownload afterwards.**

Posted snippets for `compute_checksum.sh` and `tar_and_hash_datasets.sh` above.

# Checklist:

- [x] I have performed manual end-to-end testing of the feature in my environment.
- [ ] I have added Docstrings and comments to the code.
- [ ] I have made changes to existing documentation where needed.
- [ ] I have added tests that show that the PR is functional.
- [ ] New and existing unit tests pass locally with my changes.
- [ ] I have added relevant collaborators to review the PR before merge.

### CODE REVIEW

**Comment by kavyasrinet** [comment]
Thanks for creating the proposal. I am onboard with the approach proposed here.
I completely agree with @Rebecca-Qian's review above. In addition, we should  also have datasets_folder_checksum and models_checksum just so we know which version of the model supports which dataset.

**Review by snyxan — APPROVED** [review]
This is great!!

Since we are going to have versioned dataset (and model as follow-up) for each commit, we can leverage github workflow to clean up orphan dataset/model in S3. I can help with that.

File: tools/data_scripts/default_checksums/datasets.txt [hunk 1]
```diff
@@ -0,0 +1 @@
+6055033e5dae3bdd424c132ce56d16f159fed5
```
[the remark this answers is not in the document]
Comment (anuragprat1k) [reply tools/data_scripts/default_checksums/datasets.txt:1]: Sure, I'll add a readme. Note that these hashes won't live in S3 but just in the repo (as explained in the PR notes above).

**Comment by anuragprat1k** [comment]
> Thanks for this PR!
> 
> Is it possible to extend this to models? This is super helpful not just for regressions but also with keeping models and datasets in sync with versioned containers.
> 
> Assuming that you've tested data scripts locally with these changes on linux/mac.
> 
> Good stuff!

Yes, we can extend this to models like this, or come up with something more if we want a model-dataset compatibility check, which seems to be the case from all the comments I see here. Will leave that up to you all.

**Comment by kavyasrinet** [comment]
Thanks so much! Can we get convergence on this soon ? Need to push out a dataset update pretty soon and we can use this approach to test the next push

**Review by kavyasrinet — APPROVED** [review]

**Pull request merged**